### PR TITLE
Merge the folk work to hint-learning branch

### PR DESCRIPTION
**Pull request by ChenchengLiang**
Merge to hint-learning branch

### CODE REVIEW

**Comment by pruemmer** [comment]
Thanks for the pr; I'm getting a build error however, can you add a correct reference to this library?

[info] Resolving play#play_2.10;2.1.0 ...
[warn] 	module not found: play#play_2.10;2.1.0
[warn] ==== local: tried
[warn]   /home/philipp/.ivy2/local/play/play_2.10/2.1.0/ivys/ivy.xml
[warn] ==== public: tried
[warn]   https://repo1.maven.org/maven2/play/play_2.10/2.1.0/play_2.10-2.1.0.pom
[warn] ==== uuverifiers: tried
[warn]   http://logicrunch.research.it.uu.se/maven/play/play_2.10/2.1.0/play_2.10-2.1.0.pom

**Comment by ChenchengLiang** [comment]
> Thanks for the pr; I'm getting a build error however, can you add a correct reference to this library?
> 
> [info] Resolving play#play_2.10;2.1.0 ...
> [warn] module not found: play#play_2.10;2.1.0
> [warn] ==== local: tried
> [warn] /home/philipp/.ivy2/local/play/play_2.10/2.1.0/ivys/ivy.xml
> [warn] ==== public: tried
> [warn] https://repo1.maven.org/maven2/play/play_2.10/2.1.0/play_2.10-2.1.0.pom
> [warn] ==== uuverifiers: tried
> [warn] http://logicrunch.research.it.uu.se/maven/play/play_2.10/2.1.0/play_2.10-2.1.0.pom

Yes. I added new lines in build.sbt in commitment 186ef66. I will now compile.

**Comment by ChenchengLiang** [comment]
Thanks for the pr; I'm getting a build error however, can you add a correct reference to this library?

[info] Resolving play#play_2.10;2.1.0 ...
[warn] module not found: play#play_2.10;2.1.0
[warn] ==== local: tried
[warn] /home/philipp/.ivy2/local/play/play_2.10/2.1.0/ivys/ivy.xml
[warn] ==== public: tried
[warn] https://repo1.maven.org/maven2/play/play_2.10/2.1.0/play_2.10-2.1.0.pom
[warn] ==== uuverifiers: tried
[warn] http://logicrunch.research.it.uu.se/maven/play/play_2.10/2.1.0/play_2.10-2.1.0.pom

Yes. I modified build.sbt again in commitment 71f8c8e. It will now build correctly now.

File: src/lazabs/horn/preprocessor/BooleanClauseSplitter.scala [hunk 1]
```diff
@@ -37,7 +37,7 @@ import lazabs.horn.parser.HornReader
```
Comment (pruemmer) [inline src/lazabs/horn/preprocessor/BooleanClauseSplitter.scala:37]: here you seem to have an old version of the source file?

File: src/lazabs/horn/parser/HornReader.scala [hunk 2]
```diff
@@ -1,5 +1,5 @@
 /**
```
Comment (pruemmer) [inline src/lazabs/horn/parser/HornReader.scala:1]: here you seem to have an old version of the source file?

**Review by pruemmer — CHANGES_REQUESTED** [review]
Hi Chencheng,
here are a few comments on the pr. I went through the files, and saw that your fork sometimes seems to be based on outdated versions; maybe you need to merge with the master first?
Philipp

File: build.sbt [hunk 3]
```diff
@@ -5,7 +5,7 @@ lazy val commonSettings = Seq(
     version := "2.0.4",
     homepage := Some(url("https://github.com/uuverifiers/eldarica")),
     licenses := Seq("BSD License 2.0" -> url("https://github.com/uuverifiers/eldarica/blob/master/LICENSE")),
-    scalaVersion := "2.11.12",
+    //scalaVersion := "2.11.12",
```
[the remark this answers is not in the document]
Comment (ChenchengLiang) [reply build.sbt]: Yes. Now the version is specified to 2.11.12

File: build.sbt [hunk 4]
```diff
@@ -140,9 +140,20 @@ lazy val root = (project in file(".")).
...
+    //libraryDependencies += "org.scalatest" %% "scalatest" % "3.0.5"
+    libraryDependencies +=  "org.apache.commons" % "commons-lang3" % "3.5"
+    //libraryDependencies += "play" % "play_2.10" % "2.1.0"
+    libraryDependencies += "com.typesafe.play" %% "play-json" % "2.8.1"
+    
+
+	assemblyMergeStrategy in assembly := {
+	 case PathList("META-INF", xs @ _*) => MergeStrategy.discard
+	 case x => MergeStrategy.last
+	}
```
[the remark this answers is not in the document]
Comment (ChenchengLiang) [reply build.sbt]: They are moved to lazy val root

File: src/lazabs/ServerMain.scala [hunk 5]
```diff
@@ -1,5 +1,5 @@
 /**
- * Copyright (c) 2011-2019 Philipp Ruemmer. All rights reserved.
+ * Copyright (c) 2011-2017 Philipp Ruemmer. All rights reserved.
```
[the remark this answers is not in the document]
Comment (ChenchengLiang) [reply src/lazabs/ServerMain.scala]: Now it is synchronized to master

File: src/lazabs/horn/bottomup/Main-ADT.scala [hunk 6]
```diff
@@ -0,0 +1,77 @@
+/**
```
[the remark this answers is not in the document]
Comment (ChenchengLiang) [reply src/lazabs/horn/bottomup/Main-ADT.scala]: Yes. Now they are synchronized to master

File: src/lazabs/horn/parser/HornReader.scala (hunk 2 above)
Comment (ChenchengLiang) [reply src/lazabs/horn/parser/HornReader.scala]: Yes. Now they are synchronized to master

File: src/lazabs/horn/preprocessor/BooleanClauseSplitter.scala (hunk 1 above)
Comment (ChenchengLiang) [reply src/lazabs/horn/preprocessor/BooleanClauseSplitter.scala]: Yes. Now they are synchronized to master

**Review by pruemmer — APPROVED** [review]

**Comment by pruemmer** [comment]
Thanks! I'm afraid I have some more comments:

* Could you make the class StaticAbstractionBuilderSmtHintsSelection a subclass of StaticAbstractionBuilder?
* Add your name in the preamble of files you added/changed
* Make sure that the regression tests pass. At the moment all seem to fail because your branch produces some debugging output by default. Such output should only happen when the right option is given?

**Pull request closed without merging**